### PR TITLE
MAINT: 1.14.0rc2 backports

### PR DESCRIPTION
**Pull request by tylerjereddy**
A round of backports for SciPy `1.14.0rc2` is added here. I'm skipping the CI for now and adding a TODO list below, though I did check locally on ARM mac that doc build was clean and regular testsuite passed.

Backports included so far (merge conflict resolution done manually by me noted inline):

1. gh-20841 (backport already merged prior)
2. gh-17348
3. gh-20734
4. gh-20820
5. gh-20828
6. gh-20858
7. gh-20870 (**merge conflict resolved**, related to wheel builds--caution)
8. gh-20881
9. gh-20890
10. gh-20914
11. gh-20919
12. gh-20920
13. gh-20936
14. reversion of gh-18926, accomplished via `git revert -m 1 7febfdd`
15. gh-20945
16. gh-20937 (**merge conflicts resolved** per https://github.com/scipy/scipy/pull/20937/files#r1640122369)
17. gh-20952
18. gh-20953
19. gh-20927
20. gh-20923

TODO:

- [ ] Tyler read diff carefully
- [x] need to test wheel builds after gradually turning CI back on, especially since Ralf needed to yank RC1 broken MacOS wheels
- [x] update the release notes/lists when we have all the backports we want
- [x] more backports needed:
  - ~https://github.com/scipy/scipy/pull/20920~
  - ~need resolution for https://github.com/scipy/scipy/issues/20840~
  - optional/would be nice to resolve: https://github.com/scipy/scipy/issues/19572
  - ~resolution for: https://github.com/scipy/scipy/issues/20938~
  - ~resolution for: https://github.com/scipy/scipy/issues/20943~

### CODE REVIEW

File: ci/cirrus_wheels.yml [hunk 1]
```diff
@@ -29,6 +29,8 @@ cirrus_wheels_linux_aarch64_task:
         CIBW_BUILD: cp311-manylinux* cp312-manylinux*
   env:
     CIBW_PRERELEASE_PYTHONS: False
+    CIBW_ENVIRONMENT: >
+      PKG_CONFIG_PATH=/project
```
Comment (tylerjereddy) [inline ci/cirrus_wheels.yml:33]: I think in `main` we still have the pre-release NumPy stuff here, but when resolving the merge conflict I assumed that this is the one extra thing I needed to keep in this file. I guess we'll find out for sure when we probe the wheel builds here.

**Comment by tylerjereddy** [comment]
We're up to date on the backports added here, there are no remaining known issues I'm waiting to have resolved (that are absolute blockers in any case), and the test suite + doc build/relnotes looks good locally, so I'm starting off by flushing the regular CI and if that passes I'll probe the wheel builds after.

**Comment by lucascolley** [comment]
hopefully the smoke-docs failures can be resolved by backporting gh-20927

**Comment by tylerjereddy** [comment]
@ev-br before I try to dig on it, does the `smoke-docs` set of failures like `ERROR scipy/stats/mstats.py - TypeError: warnings_errors_and_rng() missing 1 required positional argument: 'test'` ring any bells? Does something need to be pinned on the release branch? If the source of the problem isn't obvious to you, then I'll start digging.

**Comment by tylerjereddy** [comment]
yargh, ok I'll try that..

**Comment by ev-br** [comment]
Yeah, I have an idea about it. Will take a look.

**Comment by ev-br** [comment]
Actually, that's exactly what Lucas said. Or pin scipy-doctest==1.1. (Backport's pteferable)

**Comment by lucascolley** [comment]
1 (?) more 🙃 gh-20923

**Comment by tylerjereddy** [comment]
On the backport branch `python dev.py -n smoke-docs` locally produces the following after everything in `requirements/all.txt` has been installed (MacOS ARM, Python `3.12`):

```
ERROR: usage: dev.py [options] [file_or_dir] [file_or_dir] [...]
dev.py: error: unrecognized arguments: --doctest-collect=api
  inifile: /Users/treddy/github_projects/scipy/pytest.ini
  rootdir: /Users/treddy/github_projects/scipy

```

**Comment by tylerjereddy** [comment]
Why is `scipy-doctest` not in a `requirements` file or metadata of some sort?

**Comment by tylerjereddy** [comment]
Anyway, I pulled in that commit as well. Note that `python dev.py -n smoke-docs` still doesn't pass locally, though IIRC guaranteeing a pass on all platforms with all linalg backends (etc.) was never a goal I suppose.

**Comment by ev-br** [comment]
>python dev.py -n smoke-docs still doesn't pass locally

Hmmm. Does it block you now? it's early morning local time, so unless it's pressing, I'd take a look tomorrow. Mind dumping the log of failures you see somewhere? 

> Why is scipy-doctest not in a requirements file or metadata of some sort?

I always thought of this as an optional dependency. Am open to suggestions on how to be express that. --- more precisely, am going to go with whatever you say makes sense :-). 

One thing to keep in mind is, numpy had to remove it from requirements.txt because it scipy-test itself needs numpy, and if a platform does not have a numpy wheel, it tries building it from source. Not sure if it's relevant for scipy though.

**Comment by tylerjereddy** [comment]
It is not pressing, no. I thought `pyroject.toml` had support for optional deps we could propagate from, but that's not something I'm going to want to backport anyway. Thanks

**Comment by tylerjereddy** [comment]
Regular CI is passing now, proceeding with a full wheel build test.

**Comment by ev-br** [comment]
Am still curious what you see locally on arm64 Mac OS. When you've time of course.

**Comment by tylerjereddy** [comment]
Well, the wheel builds are all passing. Refguide is still being a pain with `python dev.py smoke-docs` now failing with `Received "killed" signal` in CircleCI. Since Evgeni and I have now discussed the current situation with refguide over in gh-20960, and everything seems "ok", I'm probably fine with ignoring this here.

@h-vetinari how are we feeling for `1.14.0rc2` soon-ish? There has been a flurry of backport activity, but I think we've got it all here now.

**Comment by h-vetinari** [comment]
> @h-vetinari how are we feeling for `1.14.0rc2` soon-ish? There has been a flurry of backport activity, but I think we've got it all here now.

The one thing we could still wait for is a validation that the revert of #18926 fixes things on windows (because in the case that it doesn't, we might want to choose another approach entirely and backport  #20519 rather than do the revert). I can start a run based on that now and report back in a few hours.

On the other hand, I'd be OK with releasing as-is as well, as the likelihood that the revert fixes things should be high AFAIU, and there have been a lot of changes already. Finally, there's still going to be minor things (e.g. ; ideally fixing `test_chiquare_data_types_attributes` from #20943), but AFAICT these should all be small enough to not require an rc3 (as in: we could just land them and then release) 🤞

**Comment by tylerjereddy** [comment]
I won't release tonight (US time), so if you want to trigger something and report back in 12 hours or whatever that should be fine

**Comment by h-vetinari** [comment]
I tested the revert, and can confirm that it fixes #20943 (modulo `test_chiquare_data_types_attributes` of course). So IMO good to go for rc2!

**Pull request merged**